gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/fabricservice/FabricServerImpl.java

```diff
@@ -96,39 +103,65 @@ public void start(Properties bootProperties,
   public void notifyWaiting(String regionPath,
       Set<PersistentMemberID> membersToWaitFor, Set<Integer> missingBuckets,
       PersistentMemberID myId, String message) {
-    if (GemFireXDUtils.TraceFabricServiceBoot) {
-      logger.info("Accepting WAITING notification" +
-          (message != null ? ": " + message : ""));
-    }
-    if (this.serverstatus != State.WAITING) {
-      this.previousServerStatus = this.serverstatus;
+    if (missingBuckets != null && missingBuckets.isEmpty() && membersToWaitFor.isEmpty()) {
+      if (GemFireXDUtils.TraceFabricServiceBoot) {
+        logger.info("Accepting WAITING notification" +
+            (message != null ? ": " + message : ""));
+      }
+      // only notify the FabricDataBase.postCreateDDLReplay
+      notifyTableWait(regionPath);
+    } else {
+      if (GemFireXDUtils.TraceFabricServiceBoot) {
+        logger.info("Accepting WAITING notification" +
+            (message != null ? ": " + message : ""));
+      }
+      if (this.serverstatus != State.WAITING) {
+        this.previousServerStatus = this.serverstatus;
+      }
+      this.serverstatus = State.WAITING;
+      notifyTableWait(regionPath);
+      notifyWaitingInLauncher(regionPath, membersToWaitFor, missingBuckets, myId,
+          message);
     }
-    this.serverstatus = State.WAITING;
-    notifyTableWait();
-    notifyWaitingInLauncher(regionPath, membersToWaitFor, missingBuckets, myId,
-        message);
   }
 
-  public void notifyTableInitialized() {
+  public void notifyTableInitialized(boolean initialized, String regionPath) {
     synchronized (initializationNotification) {
-      notified = true;
-      initializationNotification.notifyAll();
+      if(regionPath.equals(regionPathToWaitFor)) {
+        notified = true;
+        this.initializedOrWait = initialized;
+        initializationNotification.notifyAll();
+        regionPathToWaitFor = "";
+      }
     }
   }
 
-  private void notifyTableWait() {
+  private void notifyTableWait(String regionPath) {
     synchronized (initializationNotification) {
-      notified = true;
-      initializationNotification.notifyAll();
+      if(regionPath.equals(regionPathToWaitFor)) {
+        notified = true;
+        this.initializedOrWait = true;
+        initializationNotification.notifyAll();
+        regionPathToWaitFor = "";
+      }
     }
```

Comment:
Is the matching of region required because it can get multiple notifications when initializing parent table (or child?) for its children?

Reply:
Yes, unless all children are initialized, initialization doesn't proceed to recoverBuckets. So, parent regions return without trying to recover its buckets. Its only when isColocationComplete condition is satisfied all the bucket chain is recovered.
The notification can come for all the regions while only the last child's initialization is blocked.